lib/src/downloader.dart

```diff
@@ -1,395 +1,206 @@
+import 'dart:async';
 import 'dart:io';
+import 'dart:isolate';
 import 'dart:ui';
-import 'dart:async';
 
 import 'package:flutter/services.dart';
 import 'package:flutter/widgets.dart';
+import 'package:meta/meta.dart';
+
+part 'task.dart';
+part 'status.dart';
+part 'callback_dispatcher.dart';
+
+File _fileFromDirAndName(String dir, String name) =>
+    File('$dir${Platform.pathSeparator}$name');
 
-import 'callback_dispatcher.dart';
-import 'models.dart';
-
-///
-/// A signature function for download progress updating callback
-///
-/// * `id`: unique identifier of a download task
-/// * `status`: current status of a download task
-/// * `progress`: current progress value of a download task, the value is in
-/// range of 0 and 100
-///
-typedef void DownloadCallback(
-    String id, DownloadTaskStatus status, int progress);
-
-///
-/// A convenient class wraps all api functions of **FlutterDownloader** plugin
-///
-class FlutterDownloader {
+abstract class FlutterDownloader {
   static const _channel = const MethodChannel('vn.hunghd/downloader');
   static bool _initialized = false;
+  static var _tasksById = <String, DownloadTask>{};
 
-  static Future<Null> initialize() async {
+  static Future<void> initialize() async {
     assert(!_initialized,
-        'FlutterDownloader.initialize() must be called only once!');
+        'FlutterDownloader.initialize() must be called only once.');
 
     WidgetsFlutterBinding.ensureInitialized();
 
-    final callback = PluginUtilities.getCallbackHandle(callbackDispatcher);
-    await _channel
-        .invokeMethod('initialize', <dynamic>[callback.toRawHandle()]);
+    final callback = PluginUtilities.getCallbackHandle(dispatchCallback);
+    await _channel.invokeMethod('initialize', <dynamic>[
+      callback.toRawHandle(),
+    ]);
+
+    // Create callback listener.
+
+    final port = ReceivePort()
+      ..listen((dynamic data) {
+        final id = data[0] as String;
+        final status = data[1] as DownloadTaskStatus;
+        final progress = data[2] as double;
+        _tasksById[id]?._update(status, progress);
+      });
+    IsolateNameServer.registerPortWithName(port.sendPort, 'downloader_port');
```

Comment:
Do you think that we need a way to `removePortNameMapping ()` somewhere?

Reply:
I don't think so. `FlutterDownloader` only offers static methods, so there are no multiple intances of `FlutterDownloader` running at any time. Thus, there's never a naming conflict between ports.
Theoretically, we could add a `dispose` method, but when the app/DartVM is stopped, it automatically cleans up these resources for us, so I don't think there's a need for it.

Reply:
Ok, I think it makes sense